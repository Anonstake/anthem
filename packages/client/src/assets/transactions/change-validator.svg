<?xml version="1.0" encoding="UTF-8"?>
<svg width="42px" height="42px" viewBox="0 0 42 42" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 58 (84663) - https://sketch.com -->
    <title>change validator</title>
    <desc>Created with Sketch.</desc>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Desktop-HD" transform="translate(-359.000000, -1305.000000)">
            <g id="change-validator" transform="translate(360.000000, 1306.000000)">
                <circle id="Oval" stroke="#E1E6EC" fill="#F8F8F8" cx="20" cy="20" r="20"></circle>
                <g id="Group" transform="translate(12.000000, 12.000000)" fill="#1D818B">
                    <path d="M4.73375182,7.98160512 C4.73791106,7.98567697 4.74202695,7.98979286 4.7460988,7.99395211 L4.91331048,8.16475268 C5.22625972,8.4844188 5.22625972,8.99561348 4.91331048,9.3152796 L3.274,10.9882727 L15.6490723,10.9883338 C16.1032284,10.9883338 16.4713946,11.3565001 16.4713946,11.8106562 L16.4713946,12.069948 C16.4713946,12.5241041 16.1032284,12.8922704 15.6490723,12.8922704 L3.218,12.8912727 L4.94243397,14.6529575 C5.2553832,14.9726236 5.2553832,15.4838183 4.94243397,15.8034844 L4.77522228,15.974285 C4.45751305,16.2988133 3.93687667,16.3043412 3.61234839,15.986632 L3.60000141,15.974285 L3.60000141,15.974285 L0.28208341,12.5851539 C0.220035466,12.5217742 0.170289669,12.4508655 0.132846018,12.3754132 C-0.0527762177,12.0631412 -0.0124779257,11.6542182 0.252959925,11.3830832 L3.57087792,7.99395211 C3.88858716,7.66942382 4.40922353,7.66389589 4.73375182,7.98160512 Z" id="Combined-Shape"></path>
                    <path d="M4.73375182,0.254332396 C4.73791106,0.258404244 4.74202695,0.262520137 4.7460988,0.266679379 L4.91331048,0.437479956 C5.22625972,0.757146076 5.22625972,1.26834075 4.91331048,1.58800687 L3.274,3.261 L15.6490723,3.26106111 C16.1032284,3.26106111 16.4713946,3.62922737 16.4713946,4.08338348 L16.4713946,4.34267532 C16.4713946,4.79683142 16.1032284,5.16499768 15.6490723,5.16499768 L3.218,5.164 L4.94243397,6.92568475 C5.2553832,7.24535087 5.2553832,7.75654554 4.94243397,8.07621166 L4.77522228,8.24701224 C4.45751305,8.57154053 3.93687667,8.57706846 3.61234839,8.25935922 L3.60000141,8.24701224 L3.60000141,8.24701224 L0.28208341,4.85788118 C0.220035466,4.79450149 0.170289669,4.72359274 0.132846018,4.64814049 C-0.0527762177,4.33586844 -0.0124779257,3.92694543 0.252959925,3.65581044 L3.57087792,0.266679379 C3.88858716,-0.0578489052 4.40922353,-0.0633768375 4.73375182,0.254332396 Z" id="Combined-Shape" transform="translate(8.235697, 4.256846) rotate(-180.000000) translate(-8.235697, -4.256846) "></path>
                </g>
            </g>
        </g>
    </g>
</svg>