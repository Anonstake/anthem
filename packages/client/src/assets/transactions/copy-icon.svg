<?xml version="1.0" encoding="UTF-8"?>
<svg width="26px" height="26px" viewBox="0 0 26 26" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 58 (84663) - https://sketch.com -->
    <title>copy-icon</title>
    <desc>Created with Sketch.</desc>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Desktop-HD" transform="translate(-1367.000000, -843.000000)">
            <g id="copy-icon" transform="translate(1367.000000, 843.000000)">
                <circle id="Oval" fill="#E8F0F1" cx="13" cy="13" r="13"></circle>
                <path d="M10.5420733,10.969011 L10.5420733,13.8827809 C10.5420733,14.4518998 11.0390398,14.9202371 11.6499604,14.9202371 L14.5051438,14.9202371 L14.5051438,17.0336836 C14.5051438,17.0751819 14.4671591,17.1107518 14.426009,17.1107518 L8.02875231,17.1107518 C7.98443682,17.1107518 7.94961752,17.0751819 7.94961752,17.0336836 L7.94961752,11.0460792 C7.94961752,11.004581 7.98760222,10.969011 8.02875231,10.969011 L10.5420733,10.969011 Z M17.8921129,7.88924818 C17.9775785,7.88924818 18.0503825,7.95742387 18.0503825,8.03745621 L18.0503825,13.8827809 C18.0503825,13.9628133 17.9775785,14.030989 17.8921129,14.030989 L11.6499604,14.030989 C11.5644949,14.030989 11.4916908,13.9628133 11.4916908,13.8827809 L11.4916908,8.03745621 C11.4916908,7.95742387 11.5644949,7.88924818 11.6499604,7.88924818 L17.8921129,7.88924818 Z M17.8921129,7 L11.6499604,7 C11.0390398,7 10.5420733,7.46537321 10.5420733,8.03745621 L10.5420733,10.0797629 L8.02875231,10.0797629 C7.46214719,10.0797629 7,10.5125303 7,11.0460792 L7,17.0336836 C7,17.5672326 7.46214719,18 8.02875231,18 L14.426009,18 C14.9926141,18 15.4547613,17.5672326 15.4547613,17.0336836 L15.4547613,14.9202371 L17.8921129,14.9202371 C18.5030335,14.9202371 19,14.4518998 19,13.8827809 L19,8.03745621 C19,7.46537321 18.5030335,7 17.8921129,7 L17.8921129,7 Z" id="8H" fill="#13818C"></path>
            </g>
        </g>
    </g>
</svg>