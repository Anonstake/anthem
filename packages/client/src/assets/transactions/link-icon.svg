<?xml version="1.0" encoding="UTF-8"?>
<svg width="25px" height="25px" viewBox="0 0 25 25" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 58 (84663) - https://sketch.com -->
    <title>link-icon</title>
    <desc>Created with Sketch.</desc>
    <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <g id="Desktop-HD" transform="translate(-1473.000000, -827.000000)">
            <g id="link-icon" transform="translate(1473.000000, 827.000000)">
                <circle id="Oval" fill="#E8F0F1" cx="12.5" cy="12.5" r="12.5"></circle>
                <g id="3E" transform="translate(7.000000, 7.000000)" fill="#36939D">
                    <path d="M10.2856875,5.9296875 C10.0007188,5.9296875 9.7700625,6.1606875 9.7700625,6.4453125 C9.7700625,6.6171875 9.7700625,8.5586875 9.7686875,9.48096875 C9.76834375,9.58409375 9.768,9.898625 9.38196875,9.91134375 L1.604625,9.91134375 C1.48259375,9.91134375 1.3915,9.88796875 1.35471875,9.84671875 C1.30246875,9.78828125 1.29284375,9.6504375 1.2986875,9.5975 C1.302125,9.5734375 1.30384375,9.54903125 1.30384375,9.524625 L1.30384375,1.74728125 C1.30384375,1.64209375 1.31828125,1.5021875 1.38840625,1.43275 C1.4595625,1.36159375 1.5846875,1.36125 1.604625,1.3605625 L4.6124375,1.3605625 C4.89740625,1.3605625 5.1280625,1.1295625 5.1280625,0.8449375 C5.1280625,0.5603125 4.89740625,0.3293125 4.6124375,0.3293125 L1.63178125,0.33 C1.5750625,0.32725 1.071125,0.31178125 0.68165625,0.6813125 C0.495,0.8586875 0.27259375,1.1873125 0.27259375,1.74728125 L0.27259375,9.49746875 C0.25815625,9.65490625 0.24578125,10.142 0.57440625,10.5208125 C0.74146875,10.7133125 1.0553125,10.9425937 1.604625,10.9425937 L9.381625,10.9425937 C9.381625,10.9425937 9.3823125,10.9425937 9.38265625,10.9425937 C9.95190625,10.9425937 10.7982188,10.5538125 10.7999375,9.48234375 C10.8013125,8.5600625 10.8013125,6.6171875 10.8013125,6.4453125 C10.8013125,6.1606875 10.5703125,5.9296875 10.2856875,5.9296875" id="Fill-9"></path>
                    <path d="M10.7645313,0.6565625 C10.7638437,0.6538125 10.7635,0.65140625 10.762125,0.64865625 C10.709875,0.5218125 10.6084687,0.42040625 10.481625,0.3678125 C10.478875,0.36678125 10.4764688,0.3664375 10.4737187,0.36540625 C10.4152813,0.34271875 10.352375,0.32896875 10.2856875,0.32896875 L6.934125,0.32896875 C6.64915625,0.32896875 6.4185,0.55996875 6.4185,0.84459375 C6.4185,1.12921875 6.64915625,1.36021875 6.934125,1.36021875 L9.04096875,1.36021875 L5.40925,4.99159375 C5.2078125,5.19303125 5.2078125,5.51959375 5.40925,5.72103125 C5.50996875,5.82175 5.64196875,5.8719375 5.77396875,5.8719375 C5.90596875,5.8719375 6.037625,5.82175 6.13834375,5.72103125 L9.7700625,2.0893125 L9.7700625,4.3250625 C9.7700625,4.6096875 10.0007188,4.8406875 10.2856875,4.8406875 C10.5703125,4.8406875 10.8013125,4.6096875 10.8013125,4.3250625 L10.8013125,0.84459375 C10.8013125,0.77825 10.7875625,0.715 10.7645313,0.6565625" id="Fill-10"></path>
                </g>
            </g>
        </g>
    </g>
</svg>