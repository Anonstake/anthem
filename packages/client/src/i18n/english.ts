/** ===========================================================================
 * Primary i18n Source Catalog
 * ----------------------------------------------------------------------------
 * This is the type definition for all possible i18n text values, with
 * additional interpolated arguments if necessary. This union typed is
 * mapped to the i18n methods and the supported catalog translations to
 * type check every i18n usage in the app.
 *
 * NOTE: Running the command `yarn update-i18n` will process this file and
 * identify all the text strings by matching values between regular double
 * quotes using a regex. To use double quotes in a text string, we can use
 * the actual apostrophe characters, e.g. “my quote”, to not interfere with
 * this matching process.
 * ============================================================================
 */

/**
 * NOTE: This are the shared portfolio chart types. Some networks (e.g. Celo)
 * support custom chart views.
 */
type PORTFOLIO_CHART_TYPES =
  | "TOTAL"
  | "AVAILABLE"
  | "REWARDS"
  | "STAKING"
  | "COMMISSIONS";

export type ENGLISH =
  // LandingPage */
  | readonly ["Login"]
  | readonly ["Anthem 2019-2020. All Rights Reserved."]
  | readonly ["Earn Rewards on Cryptoassets"]
  | readonly ["LOGIN"]
  | readonly ["Connect"]
  | readonly ["Earn"]
  | readonly ["Manage"]
  | readonly ["Website"]
  | readonly ["About"]
  | readonly ["Careers"]
  | readonly ["Podcast"]
  | readonly ["Medium"]
  | readonly ["YouTube"]
  | readonly ["Telegram"]
  | readonly ["Twitter"]
  | readonly ["Discord"]
  | readonly ["Terms of Service"]
  | readonly ["Privacy Policy"]
  | readonly ["GET STARTED"]
  | readonly ["Company"]
  | readonly ["Content"]
  | readonly ["Community"]
  | readonly ["Invalid address found in URL, redirecting to login page."]

  // SideMenu */
  | readonly ["Dashboard"]
  | readonly ["Wallet"]
  | readonly ["Governance"]
  | readonly ["Settings"]
  | readonly ["Help"]
  | readonly ["Logout"]
  | readonly ["Connect an Address"]
  | readonly ["Address Connected"]
  | readonly ["Link Your Address"]
  | readonly ["LINK"]
  | readonly ["Please select a login option to begin"]
  | readonly ["Sign in with Ledger"]
  | readonly ["Sign in with address"]
  | readonly ["Change address"]
  | readonly ["Copy Address"]
  | readonly ["NETWORK:"]
  | readonly [
      "Anthem currently supports staking on Cosmos. Connect your Ledger or enter a Cosmos address to start tracking your delegations.",
    ]

  // DashboardPage */
  | readonly ["Selected"]
  | readonly ["Available"]
  | readonly ["Staking"]
  | readonly ["Rewards"]
  | readonly ["Unbonding"]
  | readonly ["Commission"]
  | readonly [PORTFOLIO_CHART_TYPES]
  | readonly ["Total"]
  | readonly ["Available"]
  | readonly ["Staking"]
  | readonly ["Rewards"]
  | readonly ["Pending"]
  | readonly ["Portfolio"]
  | readonly ["Balance"]
  | readonly ["Commissions"]
  | readonly ["Viewing"]
  | readonly ["Choose an Option..."]
  | readonly ["Please select an action first."]
  | readonly ["What do you want to do?"]
  | readonly ["Delegate"]
  | readonly ["Claim Rewards"]
  | readonly ["Error fetching balance..."]
  | readonly ["SUBMIT"]
  | readonly ["Download CSV"]
  | readonly ["Failed to download CSV data..."]
  | readonly ["No data exists yet."]
  | readonly ["Price"]
  | readonly ["Change"]
  | readonly ["Signup here."]
  | readonly ["Error fetching balance, are you sure this address exists?"]
  | readonly [
      "Please note that rewards data will not start accumulating until rewards balances are 1µatom or greater.",
    ]
  | readonly [
      "Interested in a monthly summary email about your rewards data? {{signupLink}}",
      {
        signupLink: JSX.Element;
      },
    ]

  // Chart */
  | readonly ["Unbonded"]
  | readonly ["Withdrew"]
  | readonly ["(available after 21 days)"]
  | readonly ["No ATOM balance exists yet."]
  | readonly ["No staking balance exists yet."]
  | readonly ["No commissions data exists yet."]

  // Settings Page */
  | readonly ["App Language"]
  | readonly ["Enable Dark Theme"]
  | readonly ["Disable Dark Theme"]
  | readonly ["Fiat Currency"]
  | readonly ["Choose a language"]
  | readonly ["Set your display fiat currency"]
  | readonly ["App Theme"]
  | readonly ["Loading currencies..."]
  | readonly ["Could not load currencies..."]
  | readonly ["No results."]
  | readonly ["Fiat/Crypto Currency Setting"]
  | readonly ["Email Address"]
  | readonly ["Chorus One Newsletter"]
  | readonly ["Loading..."]
  | readonly ["Signup"]
  | readonly [
      "Successfully signed up for the Chorus One newsletter. Please check your email for details.",
    ]
  | readonly [
      "Could not register your email. Is your email address typed correctly?",
    ]
  | readonly ["Monthly Rewards Summary Email"]
  | readonly ["I'm interested!"]
  | readonly ["You can signup here!"]
  | readonly [
      "An email summary delivered every month with information on the reward earnings for your address. Please enter your email below if you would be interested in receiving this in the future.",
    ]
  | readonly ["You have signed up!"]
  | readonly ["Thank you! Your interest has been recorded."]

  // Help Page */
  | readonly ["What is Anthem?"]
  | readonly [
      "Anthem is a tool to help you earn on your cryptoassets. With Anthem, you can track your portfolio and access investment opportunities. In the first version, Anthem supports staking on Cosmos. In the future, we will add other blockchains with staking, as well as other crypto-native, non-custodial investment opportunities. To stay up-to-date, follow Chorus One on {{twitterLink}} or subscribe to our {{newsletterLink}}.",
      {
        twitterLink: JSX.Element;
        newsletterLink: JSX.Element;
      },
    ]
  | readonly [" or subscribe to our "]
  | readonly ["newsletter"]
  | readonly ["How do I use Anthem?"]
  | readonly [
      "You can currently use Anthem in two ways: to track your staking investment for any Cosmos address and as a Cosmos wallet by connecting your Ledger device. Paste in your address to see important metrics, stake your Atoms, or withdraw rewards with your Ledger device. {{postLink}}",
      {
        postLink: JSX.Element;
      },
    ]
  | readonly ["You can also check out this introductory post."]
  | readonly ["What kind of data does Anthem display?"]
  | readonly [
      "Anthem currently displays transaction, balance, and reward data that is relevant to a user staking on Cosmos.",
    ]
  | readonly [
      "Your overall portfolio value adding up available, staked, unbonding, and Atoms currently in your reward, and if applicable, commissions pool.",
    ]
  | readonly [
      "Your balance that is sitting idle in your wallet. These are tokens that can immediately be accessed, e.g. to send to another address.",
    ]
  | readonly [
      "The amount of Atoms that are at stake with validators on the Cosmos Hub.",
    ]
  | readonly [
      "In the balance module, rewards show you what is currently in your reward pool. In the portfolio view, the rewards chart is plotting the rewards you earned from staking tokens over time.",
    ]
  | readonly [
      "If you've entered a validator address, information about commissions earned will be displayed. In the balance module, commissions show you the amount of commissions in that validator's pool. In the portfolio view, the commissions chart is plotting what that validator earned through commissions from his delegators.",
    ]
  | readonly [
      "Unbonding relates ATOMs that are in the process of being withdrawn from staking. These Atoms do not earn rewards anymore and will become liquid (available) once the unbonding period of currently 21 days finished.",
    ]
  | readonly ["Who should I contact with questions about Anthem?"]
  | readonly ["What will I be able to do with Anthem in the future?"]
  | readonly [
      "Anthem is a non-custodial interface to help you keep track of your holdings and interactions across blockchains, wallets, and smart contracts. Anthem will help you aggregate your data from different blockchains and wallets, display key metrics to you in a human-readable format, and help you access investment opportunities.",
    ]
  | readonly ["How can I export data from Anthem?"]
  | readonly [
      "Anthem features a handy CSV export feature that you can use for your records, e.g. to simplify your tax filing. To get your CSV report, simply click on the “Download CSV” button in the Portfolio panel after entering an address.",
    ]
  | readonly ["Who is behind Anthem?"]
  | readonly [
      "Anthem is developed by Chorus One, a startup focused on providing staking nodes and services. Please refer to {{websiteLink}} for more information.",
      {
        websiteLink: JSX.Element;
      },
    ]
  | readonly ["the Chorus One website"]
  | readonly [
      "Please join the Chorus One Community on {{telegramLink}} or {{discordLink}} to discuss about Anthem. You can also contact us via our Intercom plugin directly on the page. We are always looking for feedback, so please do not hesitate to contact us!",
      {
        telegramLink: JSX.Element;
        discordLink: JSX.Element;
      },
    ]

  // Ledger Components */
  | readonly ["Connect Your Address"]
  | readonly ["Please confirm the transaction details on your ledger device"]
  | readonly ["Ledger Connected!"]
  | readonly ["Please connect and unlock your Ledger Device..."]
  | readonly ["Sign In to Anthem"]
  | readonly ["Connect your Ledger"]
  | readonly ["Setup Delegation Transaction"]
  | readonly ["Sign Delegation Transaction"]
  | readonly ["Enter a Cosmos address to get started:"]
  | readonly ["Submit Delegation Transaction"]
  | readonly ["Delegation Transaction Submitted"]
  | readonly ["Delegation Transaction Confirmed"]
  | readonly ["Setup Rewards Claim Transaction"]
  | readonly ["Sign Rewards Claim Transaction"]
  | readonly ["Submit Rewards Claim Transaction"]
  | readonly ["Rewards Claim Transaction Submitted"]
  | readonly ["Rewards Claim Transaction Confirmed"]
  | readonly ["Connect your Ledger Device to your computer and enter your PIN."]
  | readonly ["Open the Cosmos Ledger application."]
  | readonly ["At least version v1.1.1 of Cosmos Ledger app installed."]
  | readonly ["Invalid version of the Cosmos Ledger Application found."]
  | readonly ["Please install the latest version and retry."]
  | readonly ["Retry"]
  | readonly ["Enter an Address"]
  | readonly ["Search address or transaction hash..."]
  | readonly ["Link Address"]
  | readonly ["Current address:"]
  | readonly ["Recent addresses:"]
  | readonly ["Clear all addresses"]
  | readonly ["Back"]

  // Transaction Workflow */
  | readonly ["You currently have no rewards available for withdrawal."]
  | readonly [
      "Please note that at least 1 µATOM worth of rewards is required before withdrawals can occur.",
    ]
  | readonly [
      "You have a total of {{rewards}} ({{rewardsUSD}}) rewards available to withdraw.",
      {
        rewards: JSX.Element;
        rewardsUSD: string;
      },
    ]
  | readonly ["Select the rewards you wish to withdraw from the list:"]
  | readonly ["Generate My Transaction"]
  | readonly [
      "Available balance: {{balance}} ({{balanceFiat}})",
      {
        balance: JSX.Element;
        balanceFiat: string;
      },
    ]
  | readonly ["Please enter an amount to delegate"]
  | readonly ["This network is not supported yet."]
  | readonly ["Transaction Amount (ATOM)"]
  | readonly ["Enter an amount"]
  | readonly ["Delegate All"]
  | readonly ["Default gas price: {{price}} ", { price: JSX.Element }]
  | readonly ["Enter a custom gas price."]
  | readonly ["Use the default gas price."]
  | readonly ["Note:"]
  | readonly [
      "We are currently using a default gasPrice of 0.01 µATOM, feel free to adjust this value depending on network conditions.",
    ]
  | readonly ["Gas Price (µATOM)"]
  | readonly ["Enter a gas price (µATOM)"]
  | readonly ["Gas Amount"]
  | readonly ["Enter a custom gas amount"]
  | readonly [
      "Please confirm the transaction data exactly matches what is displayed on your Ledger Device.",
    ]
  | readonly [
      "Select “Sign Transaction” to confirm the transaction details on your Ledger.",
    ]
  | readonly ["Sign Transaction"]
  | readonly ["Submitting Transaction..."]
  | readonly ["Waiting for confirmation from the blockchain..."]
  | readonly ["Transaction signed successfully!"]
  | readonly [
      "Confirm to submit your transaction to {{network}}.",
      { network: string },
    ]
  | readonly ["Cancel"]
  | readonly ["Submit"]
  | readonly ["Waiting for confirmation from the blockchain..."]
  | readonly [
      "Your transaction is successful and was included at block height {{height}}. It may take a few moments for the updates to appear in Anthem.",
      {
        height: string | number;
      },
    ]
  | readonly ["Transaction Confirmed!"]
  | readonly ["Copy Transaction Hash"]
  | readonly ["View on Mintscan"]
  | readonly ["Close"]

  // Transactions */
  | readonly ["Recent Transactions and Events"]
  | readonly ["Send"]
  | readonly ["Receive"]
  | readonly ["Vote"]
  | readonly ["Delegate"]
  | readonly ["Undelegate"]
  | readonly ["Redelegate"]
  | readonly ["Claim Rewards"]
  | readonly ["Submit Proposal"]
  | readonly ["Claim Commission"]
  | readonly ["Create Validator"]
  | readonly ["Choose Validator"]
  | readonly ["Edit Validator"]
  | readonly ["Sent"]
  | readonly ["Received"]
  | readonly ["From"]
  | readonly ["To"]
  | readonly ["Fees"]
  | readonly ["View"]
  | readonly ["Voted"]
  | readonly ["Proposal ID"]
  | readonly ["Proposer"]
  | readonly ["Title"]
  | readonly ["Description"]
  | readonly ["Delegator"]
  | readonly ["Validator"]
  | readonly ["View transaction details"]
  | readonly ["View on a block explorer"]
  | readonly ["View transaction on a block explorer"]
  | readonly ["Transaction Detail"]
  | readonly ["Transaction could not be found for hash:"]
  | readonly [
      "Warning! Transaction failed, log: {{transactionFailedLog}}",
      {
        transactionFailedLog: string;
      },
    ]

  // Toast messages */
  | readonly ["Address {{address}} copied to clipboard", { address: string }]
  | readonly ["No data found..."]
  | readonly ["Transaction hash copied."]
  | readonly [
      "Something went wrong... account information is not available right now.",
    ]
  | readonly ["The screensaver mode is currently active on the Ledger Device"]
  | readonly ["Ledger Device connected!"]
  | readonly [
      "Could not access Ledger. Is your device still connected and unlocked?",
    ]
  | readonly ["Failed to send transaction"]
  | readonly [
      "An unknown error occurred, received: {{errorString}}. Please refresh and try again.",
      { errorString: string },
    ]
  | readonly ["Logout Success."]
  | readonly ["Signing in with Ledger is currently not supported on mobile."]

  // Error Messages */
  | readonly ["This address is already connected."]
  | readonly ["This does not appear to be a valid address."]
  | readonly ["Please enter a value."]
  | readonly [
      "This appears to be a validator address. Please use the associated delegator address if you wish to view information on a validator.",
    ]
  | readonly ["Please input an amount."]
  | readonly ["Final value is greater than the maximum available."]
  | readonly ["Final amount must be greater than zero."]

  // Miscellaneous */
  | readonly ["Error fetching data..."]
  | readonly ["Are you sure you want to logout?"]
  | readonly ["Any addresses you have entered will be cleared."]
  | readonly ["Anthem encountered an error..."]
  | readonly ["Don't worry, our engineers will be on it right away!"]
  | readonly ["In the meantime you can try to reload the page."];
