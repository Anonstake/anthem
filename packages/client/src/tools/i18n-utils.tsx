import {
  catalogs,
  DEFAULT_LOCALE,
  ICatalog,
  ILocale,
  TODO_IMPLEMENTATION_MSG,
} from "i18n/catalog";
import { ENGLISH } from "i18n/english";
import ENV from "lib/client-env";
import React from "react";

/** ===========================================================================
 * Types & Config
 * ============================================================================
 */

export type tFn = (...text: ENGLISH) => ReadonlyArray<JSX.Element>;
export type tFnString = (...text: ENGLISH) => string;
export type tReturnType = ReadonlyArray<JSX.Element> | string;

export interface TranslateMethodProps {
  t: tFn;
  tString: tFnString;
}

interface I18nTranslateProps extends TranslateMethodProps {
  locale: ILocale;
}

export interface I18nProps {
  i18n: I18nTranslateProps;
  setLocale: (locale: ILocale) => void;
}

/** ===========================================================================
 * Utils
 * ============================================================================
 */

/**
 * Get a catalog given a locale if it exists, otherwise return the default
 * locale.
 */
export const getCatalogFromLocale = (locale: ILocale): ICatalog => {
  if (locale in catalogs) {
    return catalogs[locale].catalog;
  }

  return catalogs[DEFAULT_LOCALE].catalog;
};

const INTERPOLATION_PATTERN = new RegExp(/\{\{(.*?)\}\}/g);

/**
 * Helper to parse the i18n input.
 */
const parseCatalogInput = (input: ENGLISH, catalog: ICatalog) => {
  const [t, ...rest] = input;
  const variables = rest[0];

  if (!(t in catalog)) {
    if (ENV.DEVELOPMENT) {
      throw new Error(`No match for i18n text input: ${t}`);
    } else {
      // In production allow fallback to the English key.
      const text = t.split(INTERPOLATION_PATTERN);
      return {
        text,
        variables,
      };
    }
  }

  const value = catalog[t];

  if (value === TODO_IMPLEMENTATION_MSG) {
    /**
     * This shouldn't happen but if it does just fallback to the original
     * English string value.
     */
    const text = t.split(INTERPOLATION_PATTERN);
    return { text, variables };
  } else {
    const text = value.split(INTERPOLATION_PATTERN);
    return { text, variables };
  }
};

/**
 * This methods returns a translation method given a locale which interpolates
 * translations values for text internationalization.
 */
export const createTranslationMethodFromLocale = (catalog: ICatalog) => (
  ...input: ENGLISH
): ReadonlyArray<JSX.Element> => {
  const { text, variables } = parseCatalogInput(input, catalog);

  return text.filter(Boolean).map((match: string, index: number) => {
    let interpolated;
    if (variables && match in variables) {
      // index key check occurs on the line above
      interpolated = variables[match as keyof ENGLISH[1]];
    } else {
      interpolated = match;
    }

    return <span key={index}>{interpolated}</span>;
  });
};

/**
 * The same method as above but this method explicitly returns a string. This
 * is to make it easier to use the i18n-utils for components where the
 * translated text must be a string value. This method will throw an error
 * if it receives a dynamic non-string value for an interpolated variable.
 */
export const createStringTranslationMethodFromLocale = (catalog: ICatalog) => (
  ...input: ENGLISH
): string => {
  const { text, variables } = parseCatalogInput(input, catalog);

  return text
    .filter(Boolean)
    .map((match: string) => {
      let interpolated;
      if (variables && match in variables) {
        // index key check occurs on the line above
        interpolated = variables[match as keyof ENGLISH[1]];
      } else {
        interpolated = match;
      }

      if (typeof interpolated !== "string") {
        throw new Error(
          `Trying to interpolate non-string value in tString function, value: ${interpolated}`,
        );
      }

      return interpolated;
    })
    .reduce((str, itr) => str + itr, "");
};

/**
 * Helper to wrap some text string in bold attributes.
 */
export const bold = (s: string): JSX.Element => <b>{s}</b>;
