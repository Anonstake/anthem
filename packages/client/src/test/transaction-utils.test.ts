import { ITransaction } from "graphql/types";
import { ENGLISH } from "i18n/english";
import { tFn } from "tools/i18n-utils";
import { getHumanReadableMessageFromTransaction } from "tools/transaction-utils";
import transactions from "../../../utils/src/client/data/transactions.json";

const txs = transactions.transactions;

interface TxTypeMap {
  [key: string]: ReadonlyArray<ITransaction>;
}

/**
 * Transform the transactions to a map, keyed by the transaction type, to
 * be used for testing to easily retrieve transactions of a specific type.
 */
const txsByType: TxTypeMap = txs.reduce(
  // @ts-ignore
  (txMap: TxTypeMap, tx: ITransaction) => {
    const { type } = tx.msgs[0];
    let updatedMap: TxTypeMap = {};

    if (type in txMap) {
      updatedMap = {
        ...txMap,
        [type]: txMap[type].concat(tx),
      };
    } else {
      updatedMap = {
        ...txMap,
        [type]: [tx],
      };
    }

    return updatedMap;
  },
  {},
);

describe("transaction-utils", () => {
  test.skip("getHumanReadableMessageFromTransaction", () => {
    /**
     * TODO: Refactor these to test helpers.
     */
    const mockT = (s: string, vars: { [key: string]: any }) => [s, vars];
    const t = (mockT as unknown) as tFn;
    const tString = (...text: ENGLISH) => String(text);

    const txTypes: ReadonlyArray<string> = [
      "cosmos-sdk/MsgDelegate",
      "cosmos-sdk/MsgSend",
      "cosmos-sdk/MsgVote",
      "cosmos-sdk/MsgWithdrawDelegationReward",
      "cosmos-sdk/MsgWithdrawValidatorCommission",
    ];

    const testTxs = txTypes
      .map(type => {
        if (type in txsByType) {
          return txsByType[type][0];
        } else {
          return null;
        }
      })
      .filter(Boolean) as ReadonlyArray<ITransaction>;

    for (const testTx of testTxs) {
      const result = getHumanReadableMessageFromTransaction({
        t,
        tString,
        msgIndex: 0,
        denom: "uatom",
        transaction: testTx,
        address: "cosmos1gk6yv6quevfd93zwke75cn22mxhevxv00pc350",
      });
      expect(result).toMatchSnapshot();
    }
  });
});
