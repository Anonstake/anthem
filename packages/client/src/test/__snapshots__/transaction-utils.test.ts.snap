// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction-utils getHumanReadableMessageFromTransaction 1`] = `
Object {
  "data": Object {
    "amount": "2340",
    "fees": "0.109514",
    "fromAddress": "",
    "timestamp": "2019-04-28 14:51:13",
    "toAddress": "",
    "type": "custom-recieve-transcation-type",
  },
  "text": "",
}
`;

exports[`transaction-utils getHumanReadableMessageFromTransaction 2`] = `
Object {
  "data": Object {
    "amount": "142000000",
    "fees": "0.028599",
    "fromAddress": "cosmos15urq2dtp9qce4fyc85m6upwm9xul3049um7trd",
    "timestamp": "2019-05-22 14:02:48",
    "toAddress": "cosmos10whs4lsz6yjc90nzs6t4re5jv6lj59qr6zuxf8",
    "type": "custom-recieve-transcation-type",
  },
  "text": Array [
    "{{transactionType}} {{amount}} ATOM from {{recipient}} to {{sender}} with a fee of {{fee}} ATOM.",
    Object {
      "amount": <b>
        142000000
      </b>,
      "fee": <b>
        0.028599
      </b>,
      "recipient": <b>
        cosmos10whs4lsz6yjc90nzs6t4re5jv6lj59qr6zuxf8
      </b>,
      "sender": <b>
        cosmos15urq2dtp9qce4fyc85m6upwm9xul3049um7trd
      </b>,
      "transactionType": "Received",
    },
  ],
}
`;

exports[`transaction-utils getHumanReadableMessageFromTransaction 3`] = `
Object {
  "data": Object {
    "amount": "6470.110604",
    "fees": "0.077768",
    "fromAddress": "cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707",
    "timestamp": "2019-05-31 21:02:15",
    "toAddress": "cosmos15urq2dtp9qce4fyc85m6upwm9xul3049um7trd",
    "type": "cosmos-sdk/MsgWithdrawDelegationReward",
  },
  "text": Array [
    "Withdrew {{rewards}} rewards and {{commissions}} commissions from validator {{validator_address}}. Fees spent: {{fees}} ATOM",
    Object {
      "commissions": <b>
        6383.457486
      </b>,
      "fees": <b>
        0.077768
      </b>,
      "rewards": <b>
        86.653118
      </b>,
      "validator_address": <b>
        cosmosvaloper15urq2dtp9qce4fyc85m6upwm9xul3049e02707
      </b>,
    },
  ],
}
`;
